### DISCUSSION TITLE
Manual/ScriptCompilationAssemblyDefinitionFiles

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#29] [Notes]
# Manual/ScriptCompilationAssemblyDefinitionFiles

https://docs.unity3d.com/Manual/ScriptCompilationAssemblyDefinitionFiles.html

**Comment by vertxxyz** [comment]
Split each assembly into "Editor" and "Runtime" instead of littering your project with Editor folders. Packages use this layout and are a good example of proper script structure.  

When adding asmdefs to assets that have Editor folders strewn around the project, you can add one asmdef and use [Assembly Defintion References](https://docs.unity3d.com/Manual/class-AssemblyDefinitionReferenceImporter.html) for the remaining folders.

**Comment by vertxxyz** [comment]
You can use an [Assembly Definition Reference](https://docs.unity3d.com/Manual/class-AssemblyDefinitionReferenceImporter.html) to compile code into another assembly—like that of a package—giving you access to its `internal` members.
You can also create an [Assembly Definition](https://docs.unity3d.com/Manual/class-AssemblyDefinitionImporter.html) with a name matching any of the `InternalsVisibleTo` attributes mentioned in the [UnityCsReference](https://github.com/Unity-Technologies/UnityCsReference/blob/master/Editor/Mono/AssemblyInfo/AssemblyInfo.cs)[^1] (eg. `Unity.InternalAPIEditorBridge.001`) to create an assembly that has direct access to the built-in Editor assembly.
[^1]: Some of these names are in-use and may cause unexpected problems when packages using the same technique and name are in use at the same time.

**Comment by vertxxyz** [comment]
If you want to search for Assembly Definitions in your project use `t:AssemblyDefinitionAsset`.
